woocommerce/woocommerce-android
# Refactor how unseen reviews state is handled

Closes: #5752 

### Description

The way we were dealing with "unseen" state to determine if we should display a badge in specific places for unseen reviews was very complex and error prone. This PR aims at fixing several bugs and simplifying the logic: 
- Have a single source of truth to keep track of unseen notifications
- Stop relying on Event bus and its multiple event to figure out the "unread" notification count
- Avoid marking product reviews as read when the user dismisses the notifications from the status bar. We want to keep the badges visible as long as the user has not opened the new review detail screen or proactively marked all reviews as read from ReviewList screen
- seen. In light mode this spinner was invisible, so it looked like nothing happened. 
- Properly refresh the review list items to remove the unseen purple indicator when navigating back from review detail screen. 

A lot of file changes in the PR but most of them are just small changes that remove event bus usages. The main changes are: 

- Expose a new SharedFlow from `NotificationStore` in FluxC that will emit events whenever notifications are updated and this updates have an impact on "unread" state. Here are the [changes](https://github.com/wordpress-mobile/WordPress-FluxC-Android/compare/issue/5752-refactor-unseen-notification-state?expand=1)
- Add a new class [UnseenReviewsCountHandler](https://github.com/woocommerce/woocommerce-android/pull/5844/files#diff-ea25301ed2733079f224ff86b7939a24eab54acde1ec9c1302a51b00330787a1R17) that provides a refreshed unseen reviews count for product reviews, so that any part of the app that needs to keep track of this can easily subscribe and get updates. 
- Simplify [MoreMenuViewModel](https://github.com/woocommerce/woocommerce-android/pull/5844/files#diff-ebfe69d690d4fe6ecc355263ce3efe39acd4e807dd595febab79ce79fdeb7d77R26) removing any dependency on event bus.  

### Testing instructions
Play with adding new product reviews and check the following: 
- The badge is updated for the more menu tab
- The badge inside more menu screen keeps count of the unseen reviews 
- When opening review list and clicking on the "check" button a spinner is shown. If the request os successful when navigating back to more menu screen the badge should be gone
- When opening an unseen review detail and then navigating back to more menu the badge count should be decremented by one. 
- Opening review notification opens review detail screen. Navigating back to more menu will decrement the badge count by one. 
- Switching between different sites will refresh the unseen reviews badge accordingly. 

### Images/gif
Demo video of the previous scenarios tested: 


https://user-images.githubusercontent.com/2663464/154251950-921457d7-1d83-4f02-9781-dc064e4e3090.mp4
